ext/ftp/ftp.c

```diff
@@ -707,6 +707,61 @@ ftp_mlsd(ftpbuf_t *ftp, const char *path, const size_t path_len)
 }
 /* }}} */
 
+/* {{{ ftp_mlsd_parse_line
+ */
+int
+ftp_mlsd_parse_line(zval entry, const char *line) {
+
+	int linelength = strlen(line);
+	int chunklength = 0;
+	int mode = FTP_MLSD_MODE_NAME;
+
+	char name[FTP_MLSD_CHUNK_LENGTH_MAX + 1] = "";
+
+	zval zstr;
+
+	if(linelength > FTP_MLSD_LINE_LENGTH_MAX) {
+		php_error_docref(NULL, E_WARNING, "Line of entry '%s' exceed the limit of %u bytes. Skip entry.", line, FTP_MLSD_LINE_LENGTH_MAX);
+		return 1;
+	}
+
+	// Copy the line to input to not modify the original.
+	char *input = malloc(linelength);
+	strncpy(input, line, linelength);
+
+	while(strlen(input)) {
+		// get the length of the next (name or value) chunk.
+		chunklength = strcspn(input, "=;");
+		if(chunklength < 1) {
+			break;
+		}
+		if(chunklength > FTP_MLSD_CHUNK_LENGTH_MAX) {
+			php_error_docref(NULL, E_WARNING, "Chunk of entry '%s' exceed the limit of %u bytes. Skip entry.", input, FTP_MLSD_CHUNK_LENGTH_MAX);
+			return 1;
+		}
+		// Filename begins with a space.
+		if(strncmp(input, " ", 1) == 0) {
+			// Skip the space in the output.
+			input++;
+			ZVAL_STRING(&zstr, input);
+			zend_hash_str_update(Z_ARRVAL_P(&entry), "name", strlen("name"), &zstr);
+		}
+		else if(mode == FTP_MLSD_MODE_VALUE) {
+			ZVAL_STRINGL(&zstr, input, chunklength);
+			zend_hash_str_update(Z_ARRVAL_P(&entry), name, strlen(name), &zstr);
+			memset(&name, 0, FTP_MLSD_CHUNK_LENGTH_MAX);
+			mode = FTP_MLSD_MODE_NAME;
+		}
+		else if(mode == FTP_MLSD_MODE_NAME) {
+			strncpy(name, input, chunklength);
+			mode = FTP_MLSD_MODE_VALUE;
```

Comment:
Im not sure if this is better because no other function like ftp_nlist(), ftp_rawlist() or glob() return the filename as the key in the array. Mix the name and the facts seems a bit messy.